### PR TITLE
Initial Engine implementation and DB changes

### PR DESCRIPTION
**Pull request by HarveyHunt**
This is totally untested (I don't think the sort function will even work), but I'll write some tests for it later.

There is some more DB stuff to do as well as WS, I think manfred and I will send a PR together...

### CODE REVIEW

File: parking/backend/engine/alloc_engine.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+            return lots[0] if lots else None
+
+    async def commit_allocation(self, user_id: str, park_id: int) -> bool:
+        success = await self.dba.allocate_parking_lot(user_id, park_id)
+        if not success:
+            logger.warning('Failed to allocate parking lot {} to user {}'.format(park_id, user_id))
+        return success
+
+    async def remove_allocation(self, user_id: str) -> None:
+        self.user_sessions.deallocate(user_id)
```
Comment (manfredlift) [inline parking/backend/engine/alloc_engine.py:35]: Missing 'await'.

> Calling a coroutine does not start its code running – the coroutine object returned by the call doesn’t do anything until you schedule its execution.

File: parking/backend/engine/alloc_engine.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        else:
+            return lots[0] if lots else None
+
+    async def commit_allocation(self, user_id: str, park_id: int) -> bool:
+        success = await self.dba.allocate_parking_lot(user_id, park_id)
+        if not success:
+            logger.warning('Failed to allocate parking lot {} to user {}'.format(park_id, user_id))
+        return success
+
+    async def remove_allocation(self, user_id: str) -> None:
```
Comment (manfredlift) [inline parking/backend/engine/alloc_engine.py:34]: I also need a 'delete_allocation(user_id)' function to call from the Websocket handler.

File: parking/backend/engine/alloc_engine.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class AllocationEngine():
+    def __init__(self, dba: DbAccess, user_sessions: 'UserSessions') -> None:
+        self.dba = dba
+        self.user_sessions = user_sessions
+
+    async def handle_allocation_request(self, user_id: str, request: ParkingRequestMessage) -> Optional[ParkingLot]:
+        rejections = self.user_sessions.get_user(user_id).rejections
+        lots = await self.dba.get_available_parking_lots(request.location, 100, rejections)
```
Comment (manfredlift) [inline parking/backend/engine/alloc_engine.py:19]: Try to get 'distance' from ParkingRequestMessage.prefereces['distance'], if not then use a default value larger than 100 meters.

File: parking/backend/db/dbaccess.py [hunk 4]
```diff
@@ -1,8 +1,9 @@
 import logging
 from asyncio import AbstractEventLoop
-from typing import Optional
+from typing import Optional, List
 
 import asyncpg
+from asyncpg import Record
 
 import parking.backend.db.sql_constants as c
```
Comment (FlinnBurgess) [inline parking/backend/db/dbaccess.py:8]: Your use of single-letter variables is offensive to my eyes and my brain

File: parking/backend/engine/alloc_engine.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+logger = logging.getLogger('backend')
+
+
+class AllocationEngine():
+    def __init__(self, dba: DbAccess, user_sessions: 'UserSessions') -> None:
+        self.dba = dba
+        self.user_sessions = user_sessions
+
+    async def handle_allocation_request(self, user_id: str, request: ParkingRequestMessage) -> Optional[ParkingLot]:
```
Comment (manfredlift) [inline parking/backend/engine/alloc_engine.py:17]: Please return 'ParkingLotAllocation' object instead I created in ws_models. It includes the distance that should be displayed to the user:
```
@attr.s
class ParkingLotAllocation(ParkingLot):
    distance: int = attr.ib(validator=enforce_type, default=0)
```

File: parking/backend/db/dbaccess.py [hunk 6]
```diff
@@ -44,6 +46,11 @@ class DbAccess(object):
                                                p.location.longitude, p.price, p.capacity, 0)
         return park_id
 
+    async def get_parking_lot(self, park_id: int) -> Record:
+        async with self.pool.acquire() as conn:
+            parking_lot: Record = await conn.fetchrow(c.PARKINGLOTS_SELECT, park_id)
+        return parking_lot
+
     async def delete_parking_lot(self, park_id: int) -> Optional[int]:
```
Comment (FlinnBurgess) [inline parking/backend/db/dbaccess.py:54]: Can't say I like the return values of these functions. You'd expect a success or failure if you were trying to delete something, but instead you're getting either some random ID or a null value? Not only are the return types different but they also seem pretty irrelevant.

File: parking/backend/db/dbaccess.py [hunk 7]
```diff
@@ -74,3 +81,26 @@ class DbAccess(object):
                 logger.warning("Tried to allocate user : '{}' when they already had an allocation.".format(user_id))
                 return False
         return True
+
+    async def get_parking_lot_allocations(self, park_id: int) -> List[Record]:
+        async with self.pool.acquire() as conn:
+            records: List[Record] = await conn.fetch(c.ALLOCATIONS_SELECT, park_id)
+        return records
+
+    async def delete_allocation(self, user_id: str) -> None:
```
Comment (FlinnBurgess) [inline parking/backend/db/dbaccess.py:90]: Again, no indication of whether this actually works or not. How are you validating this stuff?

File: parking/backend/db/sql_constants.py [hunk 8]
```diff
@@ -63,6 +80,50 @@
...
+PARKINGLOTS_SELECT_WITHIN_DISTANCE = """
+SELECT *
+FROM (
+    SELECT
+        *, earth_distance(ll_to_earth($1, $2), ll_to_earth(lat, long)) as distance
+    FROM
+        ParkingLots
+    WHERE
+        earth_box(ll_to_earth($1, $2), $3) @> ll_to_earth(lat, long) 
+        AND num_available > num_allocated
```
Comment (FlinnBurgess) [inline parking/backend/db/sql_constants.py:103]: num_available is a bit of a misleading name, since num_allocated of them are in fact not available. Should be something like capacity, with the order of the equality reversed:
`num_allocated < capacity`
This is a lot more readable in my opinion

File: parking/backend/engine/alloc_engine.py [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+        rejections = self.user_sessions.get_user(user_id).rejections
```
Comment (FlinnBurgess) [inline parking/backend/engine/alloc_engine.py:18]: rejections is a liiiiiittle bit vague. I'm guessing it's parking lots that the user has rejected, rather than rejections of the user by the engine?

File: parking/backend/engine/alloc_engine.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+        lots = await self.dba.get_available_parking_lots(request.location, 100, rejections)
+
+        # Just take the first lot for now...
+        if lots:
+            lot = lots[0]
+            return ParkingLot(lot['capacity'], lot['name'], lot['price'], Location(lot['lat'], lot['long']))
+        else:
+            return lots[0] if lots else None
```
Comment (FlinnBurgess) [inline parking/backend/engine/alloc_engine.py]: I'm confused by this line, surely it would never get executed if lots was non-null, so what's with the check?

File: parking/backend/engine/alloc_engine.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        self.user_sessions = user_sessions
+
+    async def handle_allocation_request(self, user_id: str, request: ParkingRequestMessage) -> Optional[ParkingLot]:
+        rejections = self.user_sessions.get_user(user_id).rejections
+        lots = await self.dba.get_available_parking_lots(request.location, 100, rejections)
+
+        # Just take the first lot for now...
+        if lots:
+            lot = lots[0]
+            return ParkingLot(lot['capacity'], lot['name'], lot['price'], Location(lot['lat'], lot['long']))
```
Comment (FlinnBurgess) [inline parking/backend/engine/alloc_engine.py:24]: It would be a lot nicer if you could just pass `lot` into the constructor and get an object, rather than passing each array element in... A quick google brings up classmethods, what are your thoughts on this? 
https://stackoverflow.com/questions/141545/how-to-overload-init-method-based-on-argument-type

File: parking/backend/engine/alloc_engine.py [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+    async def remove_allocation(self, user_id: str) -> None:
+        self.user_sessions.deallocate(user_id)
+        self.dba.delete_allocation(user_id)
+        logger.info('Removed allocation for {}'.format(user_id))
+
+    async def recalculate_allocations(self, park_id: int) -> None:
+        # Deallocate users if parking spaces are taken
+        # Calculate the allocated users and find the furthest away
+        lot = await self.dba.get_parking_lot(park_id)
```
Comment (FlinnBurgess) [inline parking/backend/engine/alloc_engine.py:46]: `lot_id` seems more appropriate than `park_id`

File: parking/backend/engine/alloc_engine.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        user_sessions = self.user_sessions
+        overflow = lot.num_allocated - lot.num_available
```
Comment (FlinnBurgess) [inline parking/backend/engine/alloc_engine.py:48]: Just to reiterate on the `num_available` column name. Also, it would probably be more readable to use a variable name like `over_allocation_occured` and do the inequality on this line, then later you can do `if over_allocation_occured`

File: tests/test_db.py [hunk 14]
```diff
@@ -138,3 +139,83 @@ def teardown_module(module):
 
         assert await db.allocate_parking_lot("test_user", 1) is True
         assert await db.allocate_parking_lot("test_user", 2) is False
+
+
+@pytest.mark.asyncio
+async def test_get_available_parking_lots(event_loop):
```
Comment (FlinnBurgess) [inline tests/test_db.py:145]: I'm guessing these tests work because of your choice of location coordinates, you might want to give a brief explanation of this? It's not immediately obvious to me who doesn't know the code all that well.

File: tests/test_db.py [hunk 15]
```diff
@@ -138,3 +139,83 @@ def teardown_module(module):
...
+@pytest.mark.asyncio
+async def test_delete_allocation(event_loop):
+    with Postgresql() as postgresql:
+        db = await DbAccess.create(postgresql.url(), event_loop, reset_tables=True)
+        parking_lot = ParkingLot(100, 'test_name', 0.0, Location(0.0, 1.0089))
+        user_id = str(uuid4())
+
+        park_id = await db.insert_parking_lot(parking_lot)
+        result = await db.allocate_parking_lot(user_id, park_id)
+        assert result
```
Comment (FlinnBurgess) [inline tests/test_db.py:199]: Some of your assertions seem to be irrelevant to the name of your test. To me this suggests they should be in their own tests. You've already tested this above, so it seems pointless to put it here as well.

**Review by FlinnBurgess — COMMENTED** [review]
Perhaps a little nit-picky at times, let me know if this is helpful feedback

File: parking/backend/db/dbaccess.py (hunk 4 above)
Comment (manfredlift) [reply parking/backend/db/dbaccess.py:8]: (Written by me). This has to with preventing massive line lengths. sql_constants.py constants have really long names already and sql_constants.py is only used internally by dbaccess.py so it shouldn't be that offensive.

File: parking/backend/db/dbaccess.py (hunk 6 above)
Comment (manfredlift) [reply parking/backend/db/dbaccess.py:54]: This has to do with consistency. Since insert_parking_lot needs to return an id or nothing, we chose all other updates/deletes to return the same thing. I agree that the alternative would be true/false.

File: parking/backend/db/dbaccess.py (hunk 7 above)
Comment (manfredlift) [reply parking/backend/db/dbaccess.py:90]: There are actual tests in tests/test_db.py?

File: parking/backend/db/sql_constants.py (hunk 8 above)
Comment (manfredlift) [reply parking/backend/db/sql_constants.py:103]: In this system:
capacity = total parking spaces in a lot
num_available = number of parking spaces available as per sensors
num_allocated = number of parking spaces allocated by our system

We need all of these to make decisions and `num_allocated < capacity` wouldn't work.

File: parking/backend/engine/alloc_engine.py (hunk 9 above)
Comment (manfredlift) [reply parking/backend/engine/alloc_engine.py:18]: Yes, could add a comment or rename to user_rejections.

File: tests/test_db.py (hunk 14 above)
Comment (manfredlift) [reply tests/test_db.py:145]: This is due to 1 degree of latitude equaling to about 110km and the distance of 1 degree of longitude depends on the exact latitude. Some sanity checks and comparisons were done with some online calculators as well to choose a test case.

Comment (manfredlift) [reply tests/test_db.py:145]: @uptonfunk, this reminds me that the simulation team should be aware that we are using earth coordinates at the backend to calculate distances.

File: parking/backend/db/dbaccess.py (hunk 4 above)
Comment (FlinnBurgess) [reply parking/backend/db/dbaccess.py:8]: I can understand the reasoning behind having the sql constants and my comment doesn't really matter massively for this project, but you should write code in a way that benefits anyone who comes after you. `c` might be nice and easy for you to type but it's essentially meaningless, something like `sql_strings` would be a lot more readable. Again, not that I'm expecting you to change it!

File: parking/backend/db/dbaccess.py (hunk 6 above)
Comment (FlinnBurgess) [reply parking/backend/db/dbaccess.py:54]: Then my argument would be that this is consistently bad practice!! Wouldn't it be better to return a tuple or something with the status and the relevant data? e.g. (success = True, data = new_id) or (success = false, data = null)

File: parking/backend/db/dbaccess.py (hunk 7 above)
Comment (FlinnBurgess) [reply parking/backend/db/dbaccess.py:90]: And what if something goes wrong that is outside the scope of the unit test?

File: parking/backend/db/sql_constants.py (hunk 8 above)
Comment (FlinnBurgess) [reply parking/backend/db/sql_constants.py:103]: Okay, that makes more sense

File: parking/backend/db/dbaccess.py (hunk 6 above)
Comment (bensimner) [reply parking/backend/db/dbaccess.py:54]: Usually, in python, invalid states are flagged by raising exceptions.

If we assume nobody will try delete a parking lot that doesn't exist, then this is clearly an error and should raise an exception.

Either way, it's better to be consistent about these things than for some to be one way in one place and others to be a different way in other places.

File: parking/backend/db/dbaccess.py (hunk 4 above)
Comment (Chippiewill) [reply parking/backend/db/dbaccess.py:8]: Variable naming is always a delicate balance. I think in this instance it's ok as while `c` itself does not convey significant meaning by itself, it does in conjunction with the variables in that namespace (e.g. the purpose of `c.PARKINGLOTS_SELECT` is obvious - especially in the context it's used in)

File: parking/backend/db/dbaccess.py (hunk 6 above)
Comment (Chippiewill) [reply parking/backend/db/dbaccess.py:54]: This is again a delicate balance. For something like `delete` there are a few possibilities:

- The delete succeeds: The callee requires no further information (other than the coroutine returns)
- The delete fails because it didn't exist: The callee needs to be informed that it didn't exist
- The delete fails because of a general DB error (e.g. query took too long): The callee needs to be informed of DB errors

For the first possibility the function can just return nothing (the id is meaningless, especially as the entry no longer exists). For the third possibility we can consider this as a firmly "exceptional" circumstance so an exception should definitely be raised (which I imagine asyncpg will handle for us).

The second possibility is more nuanced. The id of the lot being deleted does not come from within the program but from external input therefore we cannot say it is "exceptional" or an error for this to occur, for instance two deletes for the same lot may come in at the same time, so one might be inclined to use a boolean return value to indicate failure. That being said the "exceptional" criteria does not apply to Python. Exceptions _are_ recommended for control flow in pythonic code (within reason) and I think in this instance would be suitable as we ultimately need to propagate this up as a 404 response.

This is basically a longwinded way of saying I agree with Ben.

File: parking/backend/db/dbaccess.py [hunk 16]
```diff
@@ -74,3 +81,26 @@ class DbAccess(object):
...
+        async with self.pool.acquire() as conn:
+            async with conn.transaction():
+                park_id: int = await conn.fetchval(c.ALLOCATIONS_DELETE, user_id)
+                if park_id:
+                    await conn.fetchval(c.PARKINGLOTS_DECREMENT_ALLOCATION, park_id)
```
Comment (Chippiewill) [inline parking/backend/db/dbaccess.py]: Big issue here as you're doing your delete and decrement in separate database transactions which can result in an inconsistent database. You should set up an explicit transaction that does both queries at once using `BEGIN`/`COMMIT`.

File: parking/backend/db/dbaccess.py [hunk 17]
```diff
@@ -74,3 +81,26 @@ class DbAccess(object):
...
+
+    async def get_available_parking_lots(self, lat: float, long: float,
+                                         dist_meters: int, exclusions: List[int]) -> List[Record]:
+        async with self.pool.acquire() as conn:
+            if exclusions:
+                print(exclusions)
```
Comment (Chippiewill) [inline parking/backend/db/dbaccess.py:101]: Debug

File: parking/backend/engine/alloc_engine.py (hunk 9 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:18]: I'll rename to user_rejections

File: parking/backend/engine/alloc_engine.py (hunk 1 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:35]: Fixed :+1:

File: parking/backend/engine/alloc_engine.py (hunk 3 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:19]: I'll check for this, it might be nice to create a model for preferences in the future.

I think that ```handle_allocation_message``` should probably raise an ```InvalidDistanceException``` that the caller will handle - thoughts?

File: parking/backend/engine/alloc_engine.py (hunk 12 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:46]: Whilst I agree, the DB stuff is using ```park_id``` so I was remaining consistent with that

File: parking/backend/engine/alloc_engine.py (hunk 13 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:48]: The ```overflow``` value gets used later though, so converting it to a bool immediately wouldn't help much. I think that ```overflow > 0``` is quite readable tbh

File: tests/test_db.py (hunk 15 above)
Comment (HarveyHunt) [reply tests/test_db.py:199]: True, I used lots of asserts to make debugging easier if the test fails in the future.

File: parking/backend/db/dbaccess.py (hunk 16 above)
Comment (HarveyHunt) [reply parking/backend/db/dbaccess.py]: The context manager (```async with conn.transaction()```) creates the transaction that the operations are performed in.

I realised that I can use ```await conn.execute(c.PARKINGLOTS_DECREMENT_ALLOCATION, park_id)``` instead of fetchval so have switched it

File: parking/backend/engine/alloc_engine.py (hunk 2 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:34]: Done :+1:

File: parking/backend/engine/alloc_engine.py (hunk 10 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py]: That's an editing error, should just be ```return None```

File: parking/backend/engine/alloc_engine.py (hunk 11 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:24]: It'd be nice to have a better way of constructing the object, a classmethod could be a good idea. I'll think about it more tomorrow and will have something better for the final PR

File: parking/backend/engine/alloc_engine.py [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+from parking.backend.db.dbaccess import DbAccess
+from parking.shared.rest_models import ParkingLot
+from parking.shared.location import Location
+from parking.shared.ws_models import ParkingRequestMessage
+
+logger = logging.getLogger('backend')
+
+
+class AllocationEngine():
+    def __init__(self, dba: DbAccess, user_sessions: 'UserSessions') -> None:
```
Comment (bensimner) [inline parking/backend/engine/alloc_engine.py:13]: Why is `UserSessions` in quotations here rather than just imported?

File: parking/backend/engine/alloc_engine.py [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+from geopy.distance import distance
+from parking.backend.db.dbaccess import DbAccess
+from parking.shared.rest_models import ParkingLot
+from parking.shared.location import Location
+from parking.shared.ws_models import ParkingRequestMessage
+
+logger = logging.getLogger('backend')
+
+
+class AllocationEngine():
```
Comment (bensimner) [inline parking/backend/engine/alloc_engine.py:12]: Should the engine handle all connections to the DbAccess? 

i.e. anything going through the REST API should go through the `AllocationEngine` ?

File: parking/backend/engine/alloc_engine.py (hunk 18 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:13]: There was a circular dependency between engine and wsserver (where UserSessions is).

@manfredlift has now split UserSessions into its own file, so this won't be an issue now.

File: parking/backend/engine/alloc_engine.py (hunk 19 above)
Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:12]: The engine is only going to be used for allocations, so I think it's fine for the REST API to communicate with the DB directly.

File: parking/backend/engine/alloc_engine.py (hunk 3 above)
Comment (bensimner) [reply parking/backend/engine/alloc_engine.py:19]: Also isn't this call inconsistent with the signature given earlier:

    async def get_available_parking_lots(self, lat: float, long: float,
                                         dist_meters: int, exclusions: List[int]) -> List[Record]:

Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:19]: Yeah, that's fixed in the newest version - we went with passing a ```Location``` object instead.

Comment (bensimner) [reply parking/backend/engine/alloc_engine.py:19]: I see, is the existence in this commit an artefact of your rebasing shenanigans by any chance

Comment (HarveyHunt) [reply parking/backend/engine/alloc_engine.py:19]: Surprisingly not ;) It's due to us discussing the interface and then forgetting it, whilst having no tests to verify if it worked. I don't know which is worse really...
